pkg/reconciler/apis/apiexportendpointsliceurls/apiexportendpointsliceurls_controller.go

```diff
@@ -0,0 +1,481 @@
+			c.enqueueAPIExportEndpointSliceByAPIBinding(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSliceByAPIBinding(obj)
+		},
+	})
+
+	_, _ = globalAPIExportEndpointSliceClusterInformer.Informer().AddEventHandler(events.WithoutSyncs(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSliceFromCache(obj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueAPIExportEndpointSliceFromCache(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSliceFromCache(obj)
+		},
+	}))
+
+	_, _ = globalAPIExportClusterInformer.Informer().AddEventHandler(events.WithoutSyncs(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlicesForAPIExport(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlicesForAPIExport(obj)
+		},
+	}))
+
+	_, _ = globalShardClusterInformer.Informer().AddEventHandler(events.WithoutSyncs(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAllAPIExportEndpointSlices(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if filterShardEvent(oldObj, newObj) {
+				c.enqueueAllAPIExportEndpointSlices(newObj)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAllAPIExportEndpointSlices(obj)
+		},
+	}))
+
+	return c, nil
+}
+
+// controller reconciles APIExportEndpointSlices. It ensures that the shard endpoints are populated
+// in the status of every APIExportEndpointSlices.
+type controller struct {
+	queue         workqueue.TypedRateLimitingInterface[string]
+	shardName     string
+	clusterClient kcpclientset.ClusterInterface
+
+	listShards                  func(selector labels.Selector) ([]*corev1alpha1.Shard, error)
+	listAPIExportEndpointSlices func() ([]*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExportEndpointSlice   func(ctx context.Context, path logicalcluster.Path, name string) (*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExport                func(path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error)
+	listAPIBindingsByAPIExport  func(apiexport *apisv1alpha1.APIExport) ([]*apisv1alpha1.APIBinding, error)
+	patchAPIExportEndpointSlice func(ctx context.Context, cluster logicalcluster.Path, patch *apisv1alpha1apply.APIExportEndpointSliceApplyConfiguration) error
+
+	apiExportEndpointSliceClusterInformer       apisv1alpha1informers.APIExportEndpointSliceClusterInformer
+	globalApiExportEndpointSliceClusterInformer apisv1alpha1informers.APIExportEndpointSliceClusterInformer
+}
+
+func (c *controller) enqueueAPIExportEndpointSliceByAPIBinding(obj interface{}) {
+	if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		obj = d.Obj
+	}
+	binding, ok := obj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("obj is supposed to be a APIBinding, but is %T", obj))
+		return
+	}
+
+	{ // local to shard
```

Comment:
why double code?

Reply:
readability. It isolates variables and makes it very clear to read two blocks. Its all about reading code and reducing mental overhead when reading the code.
// local to shard
// from cache